README.md
# useGamepadEvents

`useGamepadEvents` is a tiny, SSR-friendly hook for listening to gamepad events. It's a wrapper around the [Gamepad API](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API) designed for firing events in response to gamepad button presses, as opposed to polling for button states. It's designed to be used with [React](https://reactjs.org/) and exposes an event emitter interface.

## Installation

```bash
yarn add @beskar-labs/use-gamepad-events
```

## Usage

Here's a simple example of how to use `useGamepadEvents` to listen for 'a' button presses:

```tsx
import { useGamepadEvents } from '@beskar-labs/use-gamepad-events';

const App = () => {
  const gamepadEvents = useGamepadEvents({
    onConnect: (gamepad) => console.log(`Gamepad ${gamepad.id} connected`),
    onDisconnect: () => console.log('Gamepad disconnected'),
    onReady: () => console.log('Gamepad ready'),
  });

  gamepadEvents.on('a', () => {
    console.log('A button pressed');
  });

  return <p>Hello, world.</p>;
};
```

Here's a more complex example that uses multiple gamepad event emitters to control the `window`.

```tsx
const useGamepadNavigation = (): void => {
  const { open } = useCommandBar();
  const gamepadEvents = useGamepadEvents({
    onReady: (gamepad) => {
      window.alert(
        `${gamepad.id} connected. Press START to reload the page, SELECT to go back and use the D-Pad to navigate.`
      );
    },
  });

  gamepadEvents.on('options', window.reload);
  gamepadEvents.on('share', window.back);

  gamepadEvents.on('down', () => {
    if (typeof window === 'undefined' || open) {
      return;
    }

    window.scrollTo({ top: window.scrollY + window.innerHeight });
  });

  gamepadEvents.on('up', () => {
    if (typeof window === 'undefined' || open) {
      return;
    }

    window.scrollTo({ top: window.scrollY - window.innerHeight });
  });
};
```

## Known issues

Currently due to the amount of re-renders, it's difficult to capture a sequence of events e.g.

```tsx
const gamepadEvents = useGamepadEvents();
const [sequence, setSequence] = useState<string[]>([]);

buttons.forEach((button) => {
  gamepadEvents.on(button, () => {
    setSequence([...sequence, button]);
  });
});

console.log(sequence);
```

The sequence array is updated, but the re-renders cause the array to be reset to an empty array.

Really want to get this working to make it easier to capture something like a Konami code.

If you can think of a way to solve this, please let me know!


CHANGELOG.md
# v2.0.5 (Tue May 16 2023)

#### 🐛 Bug Fix

- Adding support for analog inputs [#51](https://github.com/beskar-co/use-gamepad-events/pull/51) ([@Flo2410](https://github.com/Flo2410))

#### Authors: 1

- Florian Hye ([@Flo2410](https://github.com/Flo2410))

---

# v2.0.4 (Sat Apr 01 2023)

#### ⚠️ Pushed to `main`

- Delete dependabot.yml ([@haydenbleasel](https://github.com/haydenbleasel))

#### Authors: 1

- Hayden Bleasel ([@haydenbleasel](https://github.com/haydenbleasel))

---

# v2.0.3 (Sat Apr 01 2023)

#### ⚠️ Pushed to `main`

- Bump deps ([@haydenbleasel](https://github.com/haydenbleasel))

#### 🔩 Dependency Updates

- Bump jest from 29.3.1 to 29.4.1 [#36](https://github.com/beskar-co/use-gamepad-events/pull/36) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint from 8.31.0 to 8.33.0 [#35](https://github.com/beskar-co/use-gamepad-events/pull/35) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @react-hookz/web from 20.0.3 to 22.0.0 [#34](https://github.com/beskar-co/use-gamepad-events/pull/34) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump typescript from 4.9.4 to 4.9.5 [#33](https://github.com/beskar-co/use-gamepad-events/pull/33) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @types/react from 18.0.26 to 18.0.27 [#32](https://github.com/beskar-co/use-gamepad-events/pull/32) ([@dependabot[bot]](https://github.com/dependabot[bot]))

#### Authors: 2

- [@dependabot[bot]](https://github.com/dependabot[bot])
- Hayden Bleasel ([@haydenbleasel](https://github.com/haydenbleasel))

---

# v1.0.6 (Sat Jan 14 2023)

#### ⚠️ Pushed to `main`

- Fix typo ([@haydenbleasel](https://github.com/haydenbleasel))
- Migrate to Beskar Labs ([@haydenbleasel](https://github.com/haydenbleasel))
- Upgrade to new Harmony ([@haydenbleasel](https://github.com/haydenbleasel))

#### 🔩 Dependency Updates

- Bump json5 from 1.0.1 to 1.0.2 [#31](https://github.com/beskar-co/use-gamepad-events/pull/31) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @types/react-dom from 18.0.8 to 18.0.10 [#22](https://github.com/beskar-co/use-gamepad-events/pull/22) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint-plugin-jest from 27.1.5 to 27.2.0 [#24](https://github.com/beskar-co/use-gamepad-events/pull/24) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @haydenbleasel/harmony from 1.1.8 to 1.1.9 [#14](https://github.com/beskar-co/use-gamepad-events/pull/14) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump tsup from 6.4.0 to 6.5.0 [#15](https://github.com/beskar-co/use-gamepad-events/pull/15) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint-plugin-react from 7.31.10 to 7.31.11 [#18](https://github.com/beskar-co/use-gamepad-events/pull/18) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint from 8.27.0 to 8.31.0 [#21](https://github.com/beskar-co/use-gamepad-events/pull/21) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump stylelint from 14.14.1 to 14.16.1 [#23](https://github.com/beskar-co/use-gamepad-events/pull/23) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump typescript from 4.8.4 to 4.9.4 [#25](https://github.com/beskar-co/use-gamepad-events/pull/25) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @types/react from 18.0.25 to 18.0.26 [#26](https://github.com/beskar-co/use-gamepad-events/pull/26) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint-plugin-n from 15.5.1 to 15.6.0 [#27](https://github.com/beskar-co/use-gamepad-events/pull/27) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @typescript-eslint/eslint-plugin from 5.42.1 to 5.47.1 [#28](https://github.com/beskar-co/use-gamepad-events/pull/28) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @react-hookz/web from 18.0.1 to 20.0.3 [#29](https://github.com/beskar-co/use-gamepad-events/pull/29) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump prettier from 2.7.1 to 2.8.1 [#30](https://github.com/beskar-co/use-gamepad-events/pull/30) ([@dependabot[bot]](https://github.com/dependabot[bot]))

#### Authors: 2

- [@dependabot[bot]](https://github.com/dependabot[bot])
- Hayden Bleasel ([@haydenbleasel](https://github.com/haydenbleasel))

---

# v1.0.5 (Fri Nov 11 2022)

#### ⚠️ Pushed to `main`

- Bump deps ([@haydenbleasel](https://github.com/haydenbleasel))

#### Authors: 1

- Hayden Bleasel ([@haydenbleasel](https://github.com/haydenbleasel))

---

# v1.0.4 (Fri Nov 11 2022)

#### ⚠️ Pushed to `main`

- Remove next dep ([@haydenbleasel](https://github.com/haydenbleasel))

#### 🔩 Dependency Updates

- Bump tsup from 6.2.3 to 6.3.0 [#3](https://github.com/haydenbleasel/use-gamepad-events/pull/3) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint-plugin-jest from 27.0.4 to 27.1.3 [#4](https://github.com/haydenbleasel/use-gamepad-events/pull/4) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @haydenbleasel/harmony from 1.0.30 to 1.0.31 [#5](https://github.com/haydenbleasel/use-gamepad-events/pull/5) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @react-hookz/web from 15.1.0 to 16.1.0 [#6](https://github.com/haydenbleasel/use-gamepad-events/pull/6) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump eslint-plugin-react from 7.31.8 to 7.31.10 [#7](https://github.com/haydenbleasel/use-gamepad-events/pull/7) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump @typescript-eslint/parser from 5.38.1 to 5.42.0 [#8](https://github.com/haydenbleasel/use-gamepad-events/pull/8) ([@dependabot[bot]](https://github.com/dependabot[bot]))
- Bump stylelint from 14.13.0 to 14.14.0 [#9](https://github.com/haydenbleasel/use-gamepad-events/pull/9) ([@dependabot[bot]](https://github.com/dependabot[bot]))

src/index.ts
/* eslint-disable id-length */
import { useEventListener } from '@react-hookz/web';
import { useCallback, useEffect, useState } from 'react';
import isEqual from 'lodash.isequal';
import debounce from 'lodash.debounce';

export type GamepadState = {
  a: boolean;
  b: boolean;
  x: boolean;
  y: boolean;
  l1: boolean;
  l2: number;
  l3: boolean;
  r1: boolean;
  r2: number;
  r3: boolean;
  share: boolean;
  options: boolean;
  up: boolean;
  down: boolean;
  left: boolean;
  right: boolean;
  back: boolean;
  leftStick: GamepadJoystick;
  rightStick: GamepadJoystick;
};

type GamepadJoystick = {
  x: number;
  y: number;
};

type UseGamepadEventsProps = {
  onConnect?: () => void;
  onDisconnect?: () => void;
  onReady?: (gamepad: Gamepad) => void;
  onLoop?: () => void;
};

const defaultState: GamepadState = {
  a: false,
  b: false,
  x: false,
  y: false,
  l1: false,
  l2: 0,
  l3: false,
  r1: false,
  r2: 0,
  r3: false,
  share: false,
  options: false,
  up: false,
  down: false,
  left: false,
  right: false,
  back: false,
  leftStick: { x: 0, y: 0 },
  rightStick: { x: 0, y: 0 },
};

const useGamepadEvents = (
  props?: UseGamepadEventsProps
): {
  on: <T extends keyof GamepadState>(
    button: T,
    callback: (value: GamepadState[T]) => void
  ) => void;
} => {
  const [gamepad, setGamepad] = useState<number | null>(null);
  const [gamepadState, setGamepadState] = useState<GamepadState>(defaultState);
  const [lastGamepadState, setlastGamepadState] =
    useState<GamepadState>(defaultState);
  const [ready, setReady] = useState(false);
  const [running, setRunning] = useState(false);
  const { onReady, onLoop, onConnect, onDisconnect } = props ?? {};

  const gameloop = useCallback(
    (oldState: GamepadState) => {
      onLoop?.();

      if (typeof gamepad !== 'number') {
        window.requestAnimationFrame(() => gameloop(oldState));
        return;
      }

      const activeGamepad = navigator.getGamepads()[gamepad];

      if (!activeGamepad) {
        window.requestAnimationFrame(() => gameloop(oldState));
        return;
      }

      const newState: GamepadState = {
        a: activeGamepad.buttons[0].pressed,
        b: activeGamepad.buttons[1].pressed,
        x: activeGamepad.buttons[2].pressed,
        y: activeGamepad.buttons[3].pressed,
        l1: activeGamepad.buttons[4].pressed,
        l2: Number(activeGamepad.buttons[6].value.toFixed(4)),
        l3: activeGamepad.buttons[10].pressed,
        r1: activeGamepad.buttons[5].pressed,
        r2: Number(activeGamepad.buttons[7].value.toFixed(4)),
        r3: activeGamepad.buttons[11].pressed,
        share: activeGamepad.buttons[8].pressed,
        options: activeGamepad.buttons[9].pressed,
        up: activeGamepad.buttons[12].pressed,
        down: activeGamepad.buttons[13].pressed,
        left: activeGamepad.buttons[14].pressed,
        right: activeGamepad.buttons[15].pressed,
        back: activeGamepad.buttons[16].pressed,
        leftStick: {
          x:
            activeGamepad.axes[0] < -0.08 || activeGamepad.axes[0] > 0.08
              ? Number(activeGamepad.axes[0].toFixed(4))
              : 0,
          y:
            activeGamepad.axes[1] < -0.08 || activeGamepad.axes[1] > 0.08
              ? Number(activeGamepad.axes[1].toFixed(4))
              : 0,
        },
        rightStick: {
          x:
            activeGamepad.axes[2] < -0.08 || activeGamepad.axes[2] > 0.08
              ? Number(activeGamepad.axes[2].toFixed(4))
              : 0,
          y:
            activeGamepad.axes[3] < -0.08 || activeGamepad.axes[3] > 0.08
              ? Number(activeGamepad.axes[3].toFixed(4))
              : 0,
        },
      };

      if (JSON.stringify(newState) !== JSON.stringify(oldState)) {
        setlastGamepadState(oldState);
        setGamepadState(newState);
      }

      const next = () => gameloop(newState);
      window.requestAnimationFrame(next);
    },
    [gamepad, onLoop]
  );

  useEffect(() => {
    if (typeof gamepad !== 'number' || ready) {
      return;
    }

    const newGamepad = navigator.getGamepads()[gamepad];

    if (!newGamepad) {
      return;
    }

    onReady?.(newGamepad);
    setReady(true);
  }, [gamepad, onReady, ready, setReady]);

  useEventListener(
    typeof window === 'undefined' ? null : window,
    'gamepadconnected',
    () => {
      const newGamepads = navigator.getGamepads();
      const activeGamePad = newGamepads.findIndex((gp) => Boolean(gp));

      setGamepad(activeGamePad);
      onConnect?.();
    },
    { passive: true }
  );

  useEventListener(
    typeof window === 'undefined' ? null : window,
    'gamepaddisconnected',
    () => {
      setGamepad(null);
      onDisconnect?.();
    },
    { passive: true }
  );

  useEffect(() => {
    if (typeof gamepad === 'number') {
      return;
    }

    const newGamepads = navigator.getGamepads();
    const activeGamePad = newGamepads.findIndex((gp) => Boolean(gp));

    if (activeGamePad === -1) {
      return;
    }

    setGamepad(activeGamePad);
  }, [gamepad]);

  useEffect(() => {
    if (typeof gamepad !== 'number' || running) {
      return;
    }

    gameloop(gamepadState);
    setRunning(true);
  }, [gameloop, gamepad, gamepadState, running]);

  const debounced_on = useCallback(
    debounce((callback: () => void) => callback(), 50, { maxWait: 50 }),
    []
  );

  const on = <T extends keyof GamepadState>(
    button: T,
    callback: (value: GamepadState[T]) => void
  ) => {
    if (isEqual(gamepadState[button], lastGamepadState[button])) return;
    debounced_on(() => callback(gamepadState[button]));
  };

  return { on };
};

export default useGamepadEvents;
